File: desktop-app (multi-platform)/desktop-app/Controlador/Host.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Net;
using System.Text;
using System.Threading.Tasks;

namespace desktop_app.Controlador {
    public class Host {
        private Socket server;
        private Socket client;

        public bool Connected {
            get {
                return this.client.Poll(1000, SelectMode.SelectError);
            }
        }

        public Host() {
            try {
                client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            } catch (Exception ex) {
                Console.WriteLine(ex.ToString() + " " + ex.Message);
            }
        }

        public Host(string ip, int port) {
            server = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            server.Bind(new IPEndPoint(IPAddress.Parse(ip), port));
            server.Listen(1);
        }

        public void Cerrar() {
            try {
                this.client?.Close();
                this.client?.Dispose();
            } catch { }
        }

        public void CerrarServer() {
            try {
                this.server?.Close();
                this.server?.Dispose();
            } catch { }
        }

        public bool Accept() {
            try {
                if (this.server.Poll(0, SelectMode.SelectRead)) {
                    this.client = this.server.Accept();
                    this.client.ReceiveTimeout = 7000;
                    this.client.SendTimeout = 7000;
                    return true;
                } else return false;
            } catch { return false; }
        }

        public bool Conectar(String ip, int port, int connectTimeout) {
            try {
                IAsyncResult ar = this.client.BeginConnect(new IPEndPoint(IPAddress.Parse(ip), port), null, null);
                if (ar.AsyncWaitHandle.WaitOne(connectTimeout)) {
                    this.client.EndConnect(ar);
                    this.client.ReceiveTimeout = 7000;
                    this.client.SendTimeout = 7000;
                    return true;
                } else { return false; }
            } catch {
                return false;
            }
        }

        public bool Enviar(byte[] data) {
            try {
                this.client.Send(data, 0, data.Length, SocketFlags.None);
                return true;
            } catch (Exception e) { Console.WriteLine(e.Message); return false; }
        }

        public bool Enviar(byte[] data, int offset, int length) {
            try {
                this.client.Send(data, offset, length, SocketFlags.None);
                return true;
            } catch (Exception e) { Console.WriteLine(e.Message); return false; }
        }

        public int Recibir(byte[] buffer) {
            try {
                return this.client.Receive(buffer, 0, buffer.Length, SocketFlags.None);
            } catch (Exception e) { Console.WriteLine(e.Message); return -1; }
        }
    }
}

File: movil-app/app/src/main/java/com/misoftware/file_sharing/Controlador/Host.java
package com.misoftware.file_sharing.Controlador;

import android.app.Application;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketException;

public class Host {
    private ServerSocket server;
    private Socket client;

    public Host() {
        this.client = new Socket();
    }

    public Host(String ip, int port) {
        try {
            this.server = new ServerSocket();
            this.server.bind(new InetSocketAddress(ip, port));
            this.server.setSoTimeout(500);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public boolean aceptar() {
        try {
            this.client = server.accept();
            this.client.setSoTimeout(7000);
            return true;
        } catch (IOException e) {
            return false;
        }
    }

    public boolean conectar(String ip, int port, int timeout) {
        try {
            this.client.connect(new InetSocketAddress(ip, port), timeout);
            return true;
        }catch (Exception e) {
            return false;
        }
    }

    public boolean enviar(byte[] data, int offset, int length) {
        try {
            OutputStream out = this.client.getOutputStream();
            out.write(data, offset, length);
            return true;
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        }
    }

    public int recibir(byte[] buffer) {
        try {
            InputStream in = this.client.getInputStream();
            return in.read(buffer, 0, buffer.length);
        } catch (IOException e) {
            e.printStackTrace();
            return -1;
        }
    }

    public void cerrar() {
        try {
            this.client.close();
        } catch (IOException e) { }
    }

    public void cerrarServer() {
        try {
            this.server.close();
        } catch (IOException e) { }
    }

    public void setSoTimeout(int timeout) {
        try {
            this.client.setSoTimeout(timeout);
        } catch (SocketException e) {
            e.printStackTrace();
        }
    }
}
